### PR TITLE
Kernel+DynamicLoader: Allow syscalls from dynamic loader more often

### PR DESCRIPTION
**Pull request by DanShaders**
Previously, if DynamicLoader were invoked directly, it wasn't able to perform syscalls after the point just before the call to the main executable entry. This made dlopen to crash if it were called from within an executable that was itself executed using `/usr/lib/Loader.so <executable>`.

### CODE REVIEW

**Comment by supercomputer7** [comment]
Commit message nitpicks:
`if DynamicLoader were invoked directly` => `if DynamicLoader was invoked directly`
`it wasn't able to perform syscalls` => `it wouldn't be able to perform syscalls`
`to crash if it were called from` => `to crash if it was called from`

Can we split the kernel change and userspace change as well into 2 separate commits?

Also, do you have an example of such program being run with `/usr/lib/Loader.so`?

File: Userland/DynamicLoader/main.cpp [hunk 1]
```diff
@@ -117,20 +100,28 @@ void _entry(int argc, char** argv, char** envp)
...
+    VERIFY(at_random_found && base_address_found);
 
     // Make sure compiler won't move any functions calls above __stack_chk_guard initialization even
     // if their definitions somehow become available.
     optimizer_fence();
 
-    perform_self_relocations(auxvp);
+    // We need to relocate ourselves.
+    // (these relocations seem to be generated because of our vtables)
+    if (!ELF::perform_relative_relocations(base_address))
```
Comment (supercomputer7) [inline Userland/DynamicLoader/main.cpp:123]: Can we put some warning message if this ever happens to fail?

**Comment by DanShaders** [comment]
Seems like was/were after "if" somehow depends on the factuality of the clause (https://ell.stackexchange.com/a/127747), though I cannot decide whether "if DynamicLoader were invoked directly" is a fact or an imaginary statement in this context.

File: Userland/DynamicLoader/main.cpp [hunk 2]
```diff
@@ -173,6 +164,20 @@ void _entry(int argc, char** argv, char** envp)
...
+
+        // Allow syscalls from our code since the kernel won't do that automatically for us if we
+        // were invoked directly.
+        Elf_Ehdr* header = reinterpret_cast<Elf_Ehdr*>(base_address);
+        Elf_Phdr* pheader = reinterpret_cast<Elf_Phdr*>(base_address + header->e_phoff);
+
+        for (size_t i = 0; i < header->e_phnum; ++i) {
+            auto const& segment = pheader[i];
+            if (segment.p_type == PT_LOAD && (segment.p_flags & PF_X)) {
+                auto flags = VirtualMemoryRangeFlags::SyscallCode | VirtualMemoryRangeFlags::Immutable;
```
[the remark this answers is not in the document]
Comment (DanShaders) [reply Userland/DynamicLoader/main.cpp:176]: This is in line with how we annotate other executable sections but I don't have any particular preferences.

**Comment by DanShaders** [comment]
> Also, do you have an example of such program being run with /usr/lib/Loader.so?

Yes! `cd /usr/Tests/LibELF; /usr/lib/Loader.so /usr/Tests/LibELF/TestDlOpen` crashes on master and works with this PR.

**Comment by supercomputer7** [comment]
> Seems like was/were after "if" somehow depends on the factuality of the clause (https://ell.stackexchange.com/a/127747), though I cannot decide whether "if DynamicLoader were invoked directly" is a fact or an imaginary statement in this context.

I don't know, to me it sounds incorrect to use `were` here, but again, it's more a nitpick than an actual problem. I see that you changed it anyway, so it looks good to me :)

**Comment by supercomputer7** [comment]
> > Also, do you have an example of such program being run with /usr/lib/Loader.so?
> 
> Yes! `cd /usr/Tests/LibELF; /usr/lib/Loader.so /usr/Tests/LibELF/TestDlOpen` crashes on master and works with this PR.

Nice! Can you include this in the commit message? I was not aware that we have such test unit, and surely by looking into the commit message one could tell how to test this easily :)

File: Userland/DynamicLoader/main.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Userland/DynamicLoader/main.cpp:178]: I see. We might as well keep it as you did it, but add a comment (a FIXME maybe?) telling to change this if we want to implement removal of the dynamic loader executable text.

[the remark this answers is not in the document]
Comment (DanShaders) [reply Userland/DynamicLoader/main.cpp:178]: Why would we want to do that? Even not considering `dl{open,close}`, DynamicLoader provides multiple runtime support functions which are impossible to implement solely in libc (see `s_magic_functions` in LibELF/DynamicLinker.cpp).

**Review by supercomputer7 — APPROVED** [review]

**Review by ADKaster — APPROVED** [review]

**Pull request merged**